gen/set.go

```diff
@@ -150,3 +151,72 @@ func (s *setGenerator) Reader(g Generator, spec *compile.SetSpec) (string, error
 
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
+
+// Equals generates a function to compare sets of the given type
+//
+// func $name(lhs, rhs $setType) bool {
+//      ...
+// }
+//
+// And returns its name.
+func (s *setGenerator) Equals(g Generator, spec *compile.SetSpec) (string, error) {
+	name := "_" + valueName(spec) + "_Equals"
+	if s.HasEquals(name) {
+		return name, nil
+	}
+
+	err := g.DeclareFromTemplate(
+		`
+            <$setType := typeReference .Spec>
+
+            <$lhs := newVar "lhs">
+            <$rhs := newVar "rhs">
+			<$o := newVar "o">
+			<$x := newVar "x">
+			<$y := newVar "y">
+			<$ok := newVar "ok">
+			func <.Name>(<$lhs>, <$rhs> <$setType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				// if the values in the set are hashable they can be used
+				// as keys in a map.
+				<if isHashable .Spec.ValueSpec>
+					<$o> := make(<$setType>, len(<$lhs>))
+					for <$x> := range <$lhs> {
+						<$o>[<$x>] = struct{}{}
+					}
```

Comment:
why are we making a copy of lhs? if the lengths of the two sets are same, all you have to check is that everything in rhs is in lhs.